relation/src/gadgets/ecc/emulated/short_weierstrass.rs

```diff
@@ -0,0 +1,406 @@
+
+        Ok(EmulatedSWPointVariable::<E>(
+            select_x,
+            select_y,
+            select_infinity,
+        ))
+    }
+
+    /// Constrain two emulated point variables to be the same.
+    /// Return error if the input point variables are invalid.
+    pub fn enforce_emulated_sw_point_equal<E: EmulationConfig<F>>(
+        &mut self,
+        p0: &EmulatedSWPointVariable<E>,
+        p1: &EmulatedSWPointVariable<E>,
+    ) -> Result<(), CircuitError> {
+        self.enforce_emulated_var_equal(&p0.0, &p1.0)?;
+        self.enforce_emulated_var_equal(&p0.1, &p1.1)?;
+        self.enforce_equal(p0.2 .0, p1.2 .0)?;
+        Ok(())
+    }
+
+    /// Obtain a bool variable representing whether two input emulated point
+    /// variables are equal. Return error if variables are invalid.
+    pub fn is_emulated_sw_point_equal<E: EmulationConfig<F>>(
+        &mut self,
+        p0: &EmulatedSWPointVariable<E>,
+        p1: &EmulatedSWPointVariable<E>,
+    ) -> Result<BoolVar, CircuitError> {
+        let mut r0 = self.is_emulated_var_equal(&p0.0, &p1.0)?;
+        let r1 = self.is_emulated_var_equal(&p0.1, &p1.1)?;
+        let r2 = self.is_equal(p0.2 .0, p1.2 .0)?;
+        r0.0 = self.mul(r0.0, r1.0)?;
+        r0.0 = self.mul(r0.0, r2.0)?;
+        Ok(r0)
+    }
+
+    /// Constrain variable `p2` to be the point addition of `p0` and
+    /// `p1` over an elliptic curve.
+    /// Let p0 = (x0, y0, inf0), p1 = (x1, y1, inf1), p2 = (x2, y2, inf2)
+    /// The addition formula for affine points of sw curve is
+    ///   If either p0 or p1 is infinity, then p2 equals to another point.
+    ///   1. if p0 == p1
+    ///     - if y0 == 0 then inf2 = 1
+    ///     - Calculate s = (3 * x0^2 + a) / (2 * y0)
+    ///     - x2 = s^2 - 2 * x0
+    ///     - y2 = s(x0 - x2) - y0
+    ///   2. Otherwise
+    ///     - if x0 == x1 then inf2 = 1
+    ///     - Calculate s = (y0 - y1) / (x0 - x1)
+    ///     - x2 = s^2 - x0 - x1
+    ///     - y2 = s(x0 - x2) - y0
+    /// The first case is equivalent to the following:
+    /// - inf0 == 1 || inf1 == 1 || x0 != x1 || y0 != y1 || y0 != 0 || inf2 == 0
+    /// - (x2 + 2 * x0) * (y0 + y0)^2 == (3 * x1^2 + a)^2
+    /// - (y2 + y0) * (y0 + y0) == (3 * x1^2 + a) (x0 - x2)
+    /// The second case is equivalent to the following:
+    /// - inf0 == 1 || inf1 == 1 || x0 != x1 || y0 == y1 || inf2 == 0
+    /// - (x0 - x1)^2 (x0 + x1 + x2) == (y0 - y1)^2
+    /// - (x0 - x2) (y0 - y1) == (y0 + y2) (x0 - x1)
+    /// First check in both cases can be combined into the following:
+    /// inf0 == 1 || inf1 == 1 || inf2 == 0 || x0 != x1 || (y0 == y1 && y0 != 0)
+    /// For the rest equality checks,
+    ///   - Both LHS and RHS must be multiplied with an indicator variable
+    ///     (!inf0 && !inf1). So that if either p0 or p1 is infinity, those
+    ///     checks will trivially pass.
+    ///   - For the first case (point doubling), both LHS and RHS must be
```

Comment:
Why not directly use a point equality check to obtain a bool variable output b? And then just multiply the bit b or 1-b?